### DISCUSSION TITLE
Does BlackBirdModule requires the parameter MethodHandles::lookup when used with Java 9+ upwards?

### DISCUSSION THREAD

**Discussion by helpermethod** [FasterXML/jackson-modules-base#264] [Q&A] [answered]
It seems with "newer" versions (Java 9 and above) Blackbird doesn't work without passing `MethodHandles::lookup` as parameter to `new BlackbirdModule()`?

The documentation isn't totally clear on that.

**Comment by cowtowncoder** [comment]
cc @stevenschlansker

**Comment by stevenschlansker** [comment, marked as answer]
Hi @helpermethod , the lookup object should not be required in classpath based deployments.
It is provided as an escape hatch in case you use modules or some other protection that requires accessing your objects with a MethodHandle with special privileges, like if it needs to come from the same module to access `private` members.

I do not pass in any lookup configuration myself when we use Blackbird.

sorry this isn't explained well in the docs. We could improve that, although I am starting to come around to the feeling that Jackson should just cut over to using MethodHandle based reflection in core and then on modern JVMs there is probably little reason to run afterburner or blackbird anymore.

Hope that helps

**Reply by cowtowncoder** [reply]
With Jackson 3.0/master having baseline of JDK 17 (at least), just change would be now doable. Fwtw.

**Reply by helpermethod** [reply]
Thanks for the answer @stevenschlansker!

This was the first time I've actually used Blackbird, for 99% of my use cases plain Jackson is fast enough.

Here is a simple reproducer:

```
#!/usr/bin/env kotlin

@file:DependsOn("com.fasterxml.jackson.module:jackson-module-kotlin:2.18.2")
@file:DependsOn("com.fasterxml.jackson.module:jackson-module-blackbird:2.18.2")

import com.fasterxml.jackson.module.blackbird.BlackbirdModule
import com.fasterxml.jackson.module.kotlin.jsonMapper

val objectMapper = jsonMapper {
    addModule(BlackbirdModule())
}

objectMapper.writeValueAsString(Person("helpermethod"))

data class Person(
    val name: String
)
```

It's a simple Kotlin script which uses Jackson w/ Blackbird to serialize a DTO. When trying to execute this code, I am getting

> java.lang.ClassFormatError: Illegal class name "/$$JacksonBlackbirdAccess" in class file /$$JacksonBlackbirdAccess
>         at java.base/java.lang.ClassLoader.defineClass0(Native Method)
>         at java.base/java.lang.System$2.defineClass(System.java:2390)
>         at java.base/java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClass(MethodHandles.java:2505)
>         at java.base/java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClass(MethodHandles.java:2480)
>         at java.base/java.lang.invoke.MethodHandles$Lookup.defineClass(MethodHandles.java:1865)
>         at com.fasterxml.jackson.module.blackbird.CrossLoaderAccess.accessClassIn(CrossLoaderAccess.java:131)
>         at com.fasterxml.jackson.module.blackbird.CrossLoaderAccess.grantAccess(CrossLoaderAccess.java:94)
>         at com.fasterxml.jackson.module.blackbird.CrossLoaderAccess.apply(CrossLoaderAccess.java:83)
>         at com.fasterxml.jackson.module.blackbird.CrossLoaderAccess.apply(CrossLoaderAccess.java:11)
>         at com.fasterxml.jackson.module.blackbird.ser.BBSerializerModifier.createProperty(BBSerializerModifier.java:111)
>         at com.fasterxml.jackson.module.blackbird.ser.BBSerializerModifier.lambda$findProperties$0(BBSerializerModifier.java:69)
>         at com.fasterxml.jackson.module.blackbird.util.Unchecked.lambda$runnable$0(Unchecked.java:31)
>         at com.fasterxml.jackson.module.blackbird.ser.BBSerializerModifier.findProperties(BBSerializerModifier.java:70)
>         at com.fasterxml.jackson.module.blackbird.ser.BBSerializerModifier.changeProperties(BBSerializerModifier.java:54)
>         at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.constructBeanOrAddOnSerializer(BeanSerializerFactory.java:415)
>         at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanOrAddOnSerializer(BeanSerializerFactory.java:295)
>         at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:240)
>         at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:174)
>         at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1523)
>         at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1471)
>         at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:578)
>         at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:856)
>         at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:330)
>         at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4811)
>         at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4052)
>         at Reproducer_main.<init>(reproducer.main.kts:13)

When I pass in MethodHandle::lookup to the BlackBirdModule constructor, it somehow works.

**Reply by stevenschlansker** [reply]
I see. That is not exactly expected - Blackbird does some magic to crack open classes in the default case, which involves defining a sibling class in the package of the target. Somehow this goes wrong with your Kotlin class. Thank you for providing the reproducer. Is providing the MethodHandles lookup workaround good enough for now? We can pursue a fix and convert this to an issue, although I cannot promise to look at a fix right away.

**Reply by cowtowncoder** [reply]
I may be totally off-base here, but would the class be in "default" package (no `package` defined)?
If so, could this be some special case for that edge case, and using non-empty package declaration would work around whatever the problem is?
